### PR TITLE
Use buffer for log writer

### LINKED ISSUE #2064: woodpecker slowed done
### Component

agent

### Describe the bug

I upgraded to the latest woodpecker "Next" version about a few weeks ago.

Since then, I notice that woodpecker is slowed down a lot (about 5-6 times slower).

I cannot really say why, but as far as I can see, it's related to the amount of console-output.

- The `php-lint` step slowed down from 0:30 to 9-10 minutes, and it's "just" static linting the code and dumping it to the output
- The `mariadb` service is starting quite fast, no delay found
- The `phpunit` step is slowed from 2 minutes to 4-5 minutes, but has not as much output as `php-lint`, but it's FAR more complex

The .woodpecker files can be found at https://github.com/friendica/friendica/blob/develop/.woodpecker/.phpunit.yml

2 month ago, it was like this:
image

now it's:
image

I moved to an exclusive, brand new root-server with 128GB ram, i7 and only docker and woodpecker to have a look if it's an IO/CPU/.. issue
--> nope the result stays the same

### System Info

```shell
woodpecker: "next-c96fc3f2"
docker info:

Client: Docker Engine - Community
 Version:    24.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.20.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 30
  Running: 17
  Paused: 0
  Stopped: 13
 Images: 22
 Server Version: 24.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local lvm
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: 856ujaye5llxyg42a78wfl9s5
  Is Manager: false
  Node Address: 10.0.3.3
  Manager Addresses:
   10.0.3.2:2377
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dcXXXXXXXXXXXXXXXXXXXXXXXX
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  userns
  cgroupns
  no-new-privileges
 Kernel Version: 6.1.0-10-amd64
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 125.7GiB
 Name: auct
 ID: 561aab6d-7adc-44a1-8f70-a7cd157e99ac
 Docker Root Dir: /var/lib/docker/165536.165536
 Debug Mode: false
 Experimental: true
 Insecure Registries:
  127.0.0.0/8


docker-compose (runner-nodes - 128GB RAM, i7, NVMe):
```console
version: '3'

services:
  default-agent:
    image: woodpeckerci/woodpecker-agent:next
    command: agent
    restart: always
    privileged: true
    userns_mode: host
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - woodpecker-default-agent-config:/etc/woodpecker
    environment:
      - WOODPECKER_SERVER=ci.friendi.ca:9000
      - WOODPECKER_AGENT_SECRET=XXXXX
      - WOODPECKER_MAX_PROCS=4
      - WOODPECKER_HOSTNAME=default-agent.opensocial.at
      - WOODPECKER_FILTER_LABELS=type=runner,location=opensocial
      - WOODPECKER_DEBUG_PRETTY=true
      - WOODPECKER_DEBUG_NOCOLOR=false

  specific-agent:
    image: woodpeckerci/woodpecker-agent:next
    command: agent
    restart: always
    privileged: true
    userns_mode: host
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - woodpecker-specific-agent-config:/etc/woodpecker
    environment:
      - WOODPECKER_SERVER=ci.friendi.ca:9000
      - WOODPECKER_AGENT_SECRET=XXXX
      - WOODPECKER_MAX_PROCS=4
      - WOODPECKER_HOSTNAME=XXX
      - WOODPECKER_FILTER_LABELS=type=runner,location=opensocial
      - WOODPECKER_DEBUG_PRETTY=true
      - WOODPECKER_DEBUG_NOCOLOR=false

volumes:
  woodpecker-default-agent-config:
  woodpecker-specific-agent-config:
```

docker-compose (server - slow node):
```console
version: '3'

services:
  server:
    image: woodpeckerci/woodpecker-server:next
    volumes:
      - woodpecker-vol-1:/var/lib/woodpecker/
    restart: always
    ports:
      - "0.0.0.0:9000:9000"
      - "0.0.0.0:8000:8000"
    extra_hosts:
      - "git.friendi.ca:192.168.173.33"
    environment:
      - WOODPECKER_OPEN=true
      - WOODPECKER_HOST=https://ci.friendi.ca
      - WOODPECKER_GITHUB=true
      - WOODPECKER_GITHUB_CLIENT=XXX
      - WOODPECKER_GITHUB_SECRET=XXX
      - WOODPECKER_AGENT_SECRET=XXXX
      - WOODPECKER_ADMIN=XXX
      - WOODPECKER_REPO_OWNERS=XXX
      - WOODPECKER_DEBUG_PRETTY=true
      - WOODPECKER_DEBUG_NOCOLOR=false
      - "WOODPECKER_DEFAULT_CLONE_IMAGE=docker.io/woodpeckerci/plugin-git:2.1.0"

  runner:
    image: woodpeckerci/woodpecker-agent:next
    restart: always
    privileged: true
    userns_mode: host
    depends_on:
      - server
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    extra_hosts:
      - "git.friendi.ca:192.168.173.33"
    environment:
      - WOODPECKER_SERVER=server:9000
      - WOODPECKER_AGENT_SECRET=XXXX
      - WOODPECKER_MAX_PROCS=2
      - WOODPECKER_HOSTNAME=runner.ci.friendi.ca
      - WOODPECKER_FILTER_LABELS=type=runner,location=friendica
      - WOODPECKER_DEBUG_PRETTY=true
      - WOODPECKER_DEBUG_NOCOLOR=false

  releaser:
    image: woodpeckerci/woodpecker-agent:next
    restart: always
    privileged: true
    hostname: releaser.woodpecker.local
    userns_mode: host
    depends_on:
      - server
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - WOODPECKER_SERVER=server:9000
      - WOODPECKER_AGENT_SECRET=XXX
      - WOODPECKER_MAX_PROCS=2
      - WOODPECKER_HOSTNAME=releaser.ci.friendi.ca
      - WOODPECKER_FILTER_LABELS=type=releaser,location=friendica
      - WOODPECKER_DEBUG_PRETTY=true
      - WOODPECKER_DEBUG_NOCOLOR=false
      - WOODPECKER_ENVIRONMENT=node:releaser

volumes:
  woodpecker-vol-1:
    external: true
    name: woodpecker_data
```
```


### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy) or the [Matrix room](https://matrix.to/#/#woodpecker:matrix.org).

### PR DESCRIPTION
**Pull request by 6543**
close  #2064

tested with a program that spils out 3000 lines of logfiles.
It took 2min on current v1.0.0./main and 4sec with this.

But the log output is broken and dont live stream anymore the bufio.Writer only consider the buffer when to write.
We alos sould consider linebreaks and time components too.

TODO:
- [x] create own bufio.Writer that:
  - [x] flushes if for a small duration no new data was received
  - [x] flushes if a specific amount of data was buffered
  - [ ] flushes on new lines before buff is full

For future PRs:
- [ ] create own bufio.Writer that:
  - [ ] flushes on new lines before buff is full
- [ ] make log save infrastructure not reliant on the unwritten promise that each write is it's own line
  - [ ] e.g. time stamp calculation#

---
*Sponsored by Kithara Software GmbH*

### CODE REVIEW

**Comment by ambroisie** [comment]
I'll just add that you should probably consider `\r` for programs that print progress bars.

**Comment by pat-s** [comment]
@6543 How far is the progress here? Is this likely to make it into 2.0?

**Comment by anbraten** [comment]
I've added a timer to the buffer so when ever a specific amount of time (currently 1s) or a specific length of data is in the buffer it is being flushed, so we should have live-logs with a maximum of a 1s delay. Added the `build_pr_images` label so everyone being interested could help testing. Code definitely needs a few improvements and tests, but I guess when everything is working it could be merged for the beginning as a first enhancement.

**Comment by xoxys** [comment]
Have these points been addressed:

> But the log output is broken and dont live stream anymore the bufio.Writer only consider the buffer when to write.
> We alos sould consider linebreaks and time components too.

**Comment by anbraten** [comment]
For the "live-streaming" the buffer will be flushed after 1 second or if it is full. Got this concept of having a 1s trigger from drone and other CI systems. Left the handling of new-lines untouched for now. Might be sth for a new PR.

File: agent/logger/buf_test.go [hunk 1]
```diff
@@ -78,16 +78,15 @@ func TestFlushAfterTime(t *testing.T) {
 		t.Fatalf("unexpected error: %v", err)
 	}
 
-	// Check if the buffer is empty
-	if len(testBuffer.buf) != 0 {
+	// check if the buffer was not flushed
```
Comment (xoxys) [inline agent/logger/buf_test.go:81]: All other comments start with an uppercase, just wondering why this one was changed?

**Comment by xoxys** [comment]
Well, that affects all other files also, e.g. `agent/logger/buf.go`. Not that I really care, but such changes look arbitrary to me.

**Comment by anbraten** [comment]
Did some testing regarding the current logging speed. Tested with a small amount of data (just the timestamps) for now.

Without the changes from this PR (unix-timestamps in ms):
- 171406337_2206 (log send by docker backend)
- 171406337_2206 (log send by agent via grpc)
- 171406337_2207 (log received by server via grpc)
- 171406337_4813 (log send by server via sse stream)
- 171406337_4815 (log received by UI)

With the changes from this PR (unix-timestamps in ms):
- 171406355_4028 (log send by docker backend)
- 171406355_4028 (log send by agent via grpc)
- 171406355_4029 (log received by server via grpc)
- 171406355_6546 (log send by server via sse stream)
- 171406355_6547 (log received by UI)

While doing some testing I noticed that our streaming logic seems to have some time issues.

**Comment by anbraten** [comment]
Generally improve the log streaming in #3722 (released in 2.6) and therefore these changes shouldn't be needed anymore.

**Pull request closed without merging**